morphman/automated_landmarking/automated_landmarking_bogunovic.py

```diff
@@ -0,0 +1,175 @@
+    """
+    Perform landmarking of an input centerline to
+    identify different segments along the vessel.
+    Landmarking algorithm based on Bogunovic (2012).
+    Bends are identified where the angle in the k1-k2 basis
+    exceeds a certain limit.
+
+    Args:
+        centerline (vtkPolyData): Centerline data points.
+        base_path (str): Location of case to landmark.
+        approximation_method (str): Method used for computing curvature.
+        algorithm (str): Name of landmarking algorithm.
+        resampling_step (float): Resampling step. Is None if no resampling.
+        smooth_line (bool): Smooths centerline with VMTK if True.
+        nknots (int): Number of knots for B-splines.
+        smoothing_factor (float): Smoothing factor used in VMTK
+        iterations (int): Number of smoothing iterations.
+        coronal_axis (str) : Axis determining coronal coordinate
+
+    Returns:
+        landmarks (dict): Landmarking interfaces as points.
+    """
+
+    if resampling_step is not None:
+        centerline = vmtk_resample_centerline(centerline, length=resampling_step)
+
+    if approximation_method == "vmtk":
+        line = vmtk_compute_centerline_attributes(centerline)
+        line = vmtk_compute_geometric_features(line, smooth_line, factor=smoothing_factor, iterations=iterations)
+        curvature = get_point_data_array("Curvature", line)
+        curvature = gaussian_filter(curvature, 2)
+
+    elif approximation_method == "disc":
+        neigh = 20
+        line = vmtk_compute_centerline_attributes(centerline)
+        line = vmtk_compute_geometric_features(line, smooth_line, factor=smoothing_factor, iterations=iterations)
+        line, curvature__ = compute_discrete_derivatives(line, neigh=neigh)
+        curvature = []
+        for c in curvature__:
+            curvature.append([c])
+        curvature = np.array(curvature)
+
+    elif approximation_method == "spline":
+        line, max_point_ids, min_point_ids = spline_centerline_and_compute_geometric_features(centerline, smooth_line,
+                                                                                              nknots)
+        curvature = get_point_data_array("Curvature", line)
+
+    if approximation_method != "spline":
+        max_point_ids = list(argrelextrema(curvature, np.greater)[0])
+        min_point_ids = list(argrelextrema(curvature, np.less)[0])
+        get_k1k2_basis(curvature, line)
+
+    length = get_curvilinear_coordinate(line)
+    k1 = get_point_data_array("k1", line)
+    k2 = get_point_data_array("k2", line)
+
+    # Remove a min / max point that is in reality a saddle point
+    for i in min_point_ids:
+        for j in max_point_ids:
+            if abs(i - j) < 5 and abs(curvature[i] - curvature[j]) < 0.01:
+                if i in min_point_ids:
+                    min_point_ids.remove(i)
+                if j in max_point_ids:
+                    max_point_ids.remove(j)
+
+    k1_points = k1[max_point_ids]
```

Comment:
Please add a comment for this block of code. Unless you read the Bogunovic paper or your paper it is tricky to understand what is going on her.